File: content/compute/lab/support/race-condition/python/race_condition.py
# SPDX-License-Identifier: BSD-3-Clause

from threading import Thread

NUM_ITER = 10_000_000
var = 0


def increment_var():
    global var
    for _ in range(NUM_ITER):
        var += 1


def decrement_var():
    global var
    for _ in range(NUM_ITER):
        var -= 1


def main():
    t1 = Thread(
        target=increment_var,
    )
    t2 = Thread(
        target=decrement_var,
    )

    t1.start()
    t2.start()

    t1.join()
    t2.join()

    print(f"var = {var}")


if __name__ == "__main__":
    exit(main())

File: content/compute/lab/support/race-condition/c/race_condition.c
// SPDX-License-Identifier: BSD-3-Clause

#include <stdio.h>
#include <pthread.h>

#include "utils/utils.h"

#define NUM_ITER 10000000

static int val;

void *increment_var(void *arg)
{
	(void)arg;

	for (size_t i = 0; i < NUM_ITER; i++)
		val++;

	return NULL;
}

void *decrement_var(void *arg)
{
	(void)arg;

	for (size_t i = 0; i < NUM_ITER; i++)
		val--;

	return NULL;
}

int main(void)
{
	int rc;
	pthread_t tids[2];

	rc = pthread_create(tids, NULL, increment_var, NULL);
	DIE(rc < 0, "pthread_create");
	rc = pthread_create(tids + 1, NULL, decrement_var, NULL);
	DIE(rc < 0, "pthread_create");

	rc = pthread_join(tids[0], NULL);
	DIE(rc < 0, "pthread_join");
	rc = pthread_join(tids[1], NULL);
	DIE(rc < 0, "pthread_join");

	printf("var = %d\n", val);

	return 0;
}
